File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/service/ContentEngagementService.kt
package com.nebras.mobile.core.service

import android.util.Log
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.SetOptions
import com.nebras.mobile.core.firebase.FirestoreSyncConfig
import com.nebras.mobile.core.model.Content
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.concurrent.ConcurrentHashMap

/**
 * عدّادات المشاهدة على Firestore — **مجهولة** (بدون UID).
 *
 * - `view_count`: عند فتح المحتوى (debounce دقيقة/عنصر).
 * - `play_count`: عند تجاوز 25% من المدّة (مرّة واحدة/جلسة).
 * - `complete_count`: عند تجاوز 90% (مرّة واحدة/جلسة).
 */
object ContentEngagementService {

    private const val TAG = "ContentEngagement"
    private const val FILES_COLLECTION = "content_unified_files"

    // 🚫 YouTube مقطوع: لا نكتب أيّ عدّاد تفاعل في `content_unified_youtube`.
    private const val VIEW_DEBOUNCE_MILLIS = 60_000L
    private const val FLUSH_EVERY_MILLIS = 5_000L

    private val db get() = FirestoreSyncConfig.instance
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val lastViewAt = ConcurrentHashMap<String, Long>()
    private val pendingViewIncrements = ConcurrentHashMap<String, Int>()
    private val playRecordedThisSession = ConcurrentHashMap.newKeySet<String>()
    private val completeRecordedThisSession = ConcurrentHashMap.newKeySet<String>()

    @Volatile
    private var flushJob: Job? = null

    fun recordContentOpened(content: Content) = recordView(content)

    fun recordView(content: Content) {
        val id = content.id.trim()
        if (id.isEmpty()) return
        // منشورات المجتمع تملك عدّاداً أسبوعياً مستقلاً في UgcService. منعها
        // هنا مهمّ حتى لا ننشئ وثائق شبحيّة في content_unified_files بنفس المعرّف.
        if (id.startsWith("ugc_")) return

        val now = System.currentTimeMillis()
        val last = lastViewAt[id]
        if (last != null && now - last < VIEW_DEBOUNCE_MILLIS) return
        lastViewAt[id] = now
        pendingViewIncrements.merge(id, 1, Int::plus)
        scheduleFlush()
    }

    fun recordPlayMilestone(content: Content, ratio: Double) {
        val id = content.id.trim()
        if (id.isEmpty()) return
        if (id.startsWith("ugc_")) return

        if (ratio >= 0.9 && completeRecordedThisSession.add(id)) {
            scope.launch {
                runCatching {
                    docRef(content).set(
                        mapOf(
                            "complete_count" to FieldValue.increment(1),
                            "last_played_at" to FieldValue.serverTimestamp(),
                        ),
                        SetOptions.merge(),
                    ).await()
                }.onFailure { Log.d(TAG, "complete_count failed: ${it.message}") }
            }
        }

        if (ratio < 0.25 || !playRecordedThisSession.add(id)) return
        scope.launch {
            runCatching {
                docRef(content).set(
                    mapOf(
                        "play_count" to FieldValue.increment(1),
                        "last_played_at" to FieldValue.serverTimestamp(),
                    ),
                    SetOptions.merge(),
                ).await()
            }.onFailure { Log.d(TAG, "play_count failed: ${it.message}") }
        }
    }

    private fun scheduleFlush() {
        if (flushJob != null) return
        flushJob = scope.launch {
            delay(FLUSH_EVERY_MILLIS)
            flushJob = null
            flushPendingViews()
        }
    }

    private suspend fun flushPendingViews() {
        if (pendingViewIncrements.isEmpty()) return
        val batch = HashMap(pendingViewIncrements)
        pendingViewIncrements.clear()

        for ((id, increment) in batch) {
            runCatching {
                // كلّ المحتوى المعروض يعيش في `content_unified_files` (مع مرآة
                // dashboard_uploads بنفس المعرّف) — YouTube مقطوع فلا مسار احتياطيّ.
                db.collection(FILES_COLLECTION).document(id).set(
                    mapOf(
                        "view_count" to FieldValue.increment(increment.toLong()),
                        "last_played_at" to FieldValue.serverTimestamp(),
                    ),
                    SetOptions.merge(),
                ).await()
            }.onFailure { Log.d(TAG, "view_count failed $id: ${it.message}") }
        }
    }

    /** 🚫 YouTube مقطوع: كلّ العدّادات تذهب لمجموعة الملفّات الموحَّدة. */
    private fun docRef(content: Content): DocumentReference =
        db.collection(FILES_COLLECTION).document(content.id)
}

File: app-flutter/lib/core/services/content_engagement_service.dart
import 'dart:async';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import 'package:nebras_mobile_app/core/firebase/firestore_sync_config.dart';
import 'package:nebras_mobile_app/features/content/model/content_model.dart';

/// عدّادات المشاهدة على Firestore — **مجهولة** (بدون UID).
///
/// - `view_count`: عند فتح المحتوى (debounce دقيقة/عنصر).
/// - `play_count`: عند تجاوز 25% من المدّة (مرّة واحدة/جلسة).
/// - `complete_count`: عند تجاوز 90% (مرّة واحدة/جلسة).
///
/// راجع [docs/CONTENT_ORDERING_AND_HOME.md].
class ContentEngagementService {
  ContentEngagementService._();

  static final ContentEngagementService instance =
      ContentEngagementService._();

  static const _filesCollection = 'content_unified_files';
  static const _youtubeCollection = 'content_unified_youtube';
  static const _viewDebounce = Duration(minutes: 1);
  static const _flushEvery = Duration(seconds: 5);

  // قاعدة Firestore المُسمّاة `default` (وليس `(default)`) — انظر FirestoreSyncConfig.
  final FirebaseFirestore _db = FirestoreSyncConfig.instance;
  final Map<String, DateTime> _lastViewAt = {};
  final Map<String, int> _pendingViewIncrements = {};
  final Set<String> _playRecordedThisSession = {};
  final Set<String> _completeRecordedThisSession = {};
  Timer? _flushTimer;

  Future<void> recordContentOpened(Content content) => recordView(content);

  Future<void> recordView(Content content) async {
    final id = content.id.trim();
    if (id.isEmpty) return;
    final now = DateTime.now();
    final last = _lastViewAt[id];
    if (last != null && now.difference(last) < _viewDebounce) return;
    _lastViewAt[id] = now;
    _pendingViewIncrements[id] = (_pendingViewIncrements[id] ?? 0) + 1;
    _scheduleFlush();
  }

  Future<void> recordPlayMilestone(Content content, {required double ratio}) async {
    final id = content.id.trim();
    if (id.isEmpty) return;

    if (ratio >= 0.9 && !_completeRecordedThisSession.contains(id)) {
      _completeRecordedThisSession.add(id);
      try {
        final ref = _docRef(content);
        await ref.set({
          'complete_count': FieldValue.increment(1),
          'last_played_at': FieldValue.serverTimestamp(),
        }, SetOptions(merge: true));
      } catch (e) {
        debugPrint('[ContentEngagement] complete_count failed: $e');
      }
    }

    if (ratio < 0.25 || _playRecordedThisSession.contains(id)) return;
    _playRecordedThisSession.add(id);
    try {
      final ref = _docRef(content);
      await ref.set({
        'play_count': FieldValue.increment(1),
        'last_played_at': FieldValue.serverTimestamp(),
      }, SetOptions(merge: true));
    } catch (e) {
      debugPrint('[ContentEngagement] play_count failed: $e');
    }
  }

  void _scheduleFlush() {
    _flushTimer ??= Timer(_flushEvery, () {
      _flushTimer = null;
      unawaited(_flushPendingViews());
    });
  }

  Future<void> _flushPendingViews() async {
    if (_pendingViewIncrements.isEmpty) return;
    final batch = Map<String, int>.from(_pendingViewIncrements);
    _pendingViewIncrements.clear();

    for (final entry in batch.entries) {
      try {
        // نحاول الملفات ثمّ YouTube — أحد المسارين يطابق المعرّف.
        final filesRef = _db.collection(_filesCollection).doc(entry.key);
        final ytRef = _db.collection(_youtubeCollection).doc(entry.key);
        final snap = await filesRef.get();
        final target = snap.exists ? filesRef : ytRef;
        await target.set({
          'view_count': FieldValue.increment(entry.value),
          'last_played_at': FieldValue.serverTimestamp(),
        }, SetOptions(merge: true));
      } catch (e) {
        debugPrint('[ContentEngagement] view_count failed ${entry.key}: $e');
      }
    }
  }

  DocumentReference<Map<String, dynamic>> _docRef(Content content) {
    if (content.type == ContentType.youtube) {
      return _db.collection(_youtubeCollection).doc(content.id);
    }
    return _db.collection(_filesCollection).doc(content.id);
  }
}
